src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -64,27 +75,66 @@ private void ComputeIndices()
             }
         }
 
+        /// <summary>
+        /// Tries to get a display order for a property context.
+        /// </summary>
+        /// <param name="propertyContext"></param>
+        /// <param name="displayOrder"></param>
+        /// <returns></returns>
+        private bool TryGetDisplayOrder(IProjectTreeCustomizablePropertyContext propertyContext, out int displayOrder)
+        {
+            // Find a display order using the item's name. This is not a relative path.
+            if (_displayOrderMap.TryGetValue(propertyContext.ItemName, out displayOrder))
+            {
+                return true;
+            }
+
+            // If we did not get a display order, we need to check its relative path.
+            // If the item has a valid Link property, use its value to find a display order.
+            var isLinked =
+                propertyContext.Metadata.TryGetValue(LinkProperty, out var linkPath) &&
+                !string.IsNullOrWhiteSpace(linkPath);
+            if (isLinked)
+            {
+                // Link path is always relative.
+                // Make sure we choose the property context with the same name as the last part of the link path.
+                if (string.Equals(propertyContext.ItemName, GetStringParts(linkPath).Last()))
```

Comment:
What happens with:

```
    <Compile Include="..\Foo.cs">
      <Link>Bar.cs</Link>
    </Compile>
```

Reply:
In this case, CPS will send a property context with an ItemName "Bar.cs" because that is how it is rendered. This same property context will also have a link path that has the same filename as it as well.

However, sometimes CPS will get into a state where it will have both "Foo.cs" and "Bar.cs" property context. This happens when you start changing a file that is not linked, into a linked a file, then back again. When we need to be careful with that because we don't want two items with the same DisplayOrder that is greater than zero.